### PR TITLE
Tommy/admin model

### PR DESCRIPTION
**Pull request by tommypoa**
## Feature Name: Admin user type
Feature Description: Admin User Model
Admin powers - when logged in as admin:
- Ability to hide works from both Works page and Artist Profile page (by hovering above work)
- Ability to hide artists - all associated works will be hidden from main page as well
- Locking and unlocking of an artist - lock and unlock button in Artist Profile page. When locked, Artist will not be able to sign in

Admin user log-in and out routes
- To log in, manually enter request in browser /admins/sign_in
- Log-out via drop-down menu

### Related PRs
None. 
### Migrations
Yes:
- Admin devise model and controllers
- Admin account_id
- :hidden field to Artists and Works table
- :lockable to Artist and Buyers

### Tests Performed, Edge Cases
- Tested on a bunch of users and admins
### Screenshots

CC: @by-co

### CODE REVIEW

**Comment by by-co** [comment]
Omg this is SO much work done, huge props Tommy!!! 🎉💪 

I have a few feature requests — can we remove the `admins/sign_up` route from rendering a signup page? Also, can we create a `/admin` route that redirects to `admins/sign_up`? Not sure if this possible with devise though 😮

File: config/routes.rb [hunk 1]
```diff
@@ -58,5 +59,8 @@
     get 'works/thumbnail/:id' => 'works#thumbnail'
     get 'artists/commissions/:id' => 'artists#commissions'
     get 'requests/request_exist/:search_params' => 'requests#request_exist'
+    put 'artists/lock_user/:id' => 'artists#lock_user'
```
Comment (by-co) [inline config/routes.rb]: What's the difference of `put`?

File: app/views/admins/registrations/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<div class="loginPage">
```
Comment (by-co) [inline app/views/admins/registrations/new.html.erb]: If it's not possible to remote the `admins/sign_up` route altogether, let's get rid of the HTML. We don't want any user to "find" the URL and make admin accounts :O

File: app/javascript/components/works/WorkColumnPanel.jsx [hunk 3]
```diff
@@ -37,6 +37,9 @@ class WorkColumnPanel extends React.Component {
           </Touchable>
         }
         <h6>{work.material}</h6>
+        {work.hidden &&
+          <h6> HIDDEN!!!!! </h6>
```
Comment (by-co) [inline app/javascript/components/works/WorkColumnPanel.jsx]: @thumn you'll be giving this a little more style at hackathon!

File: app/javascript/components/artists/ArtistProfile.jsx [hunk 4]
```diff
@@ -171,22 +256,42 @@ class ArtistProfile extends React.Component {
...
               if (this.getAvailability(activeFilter).includes(work.availability)) {
               return (
-                    <WorkColumnPanel work={work} key={work.id} hideArtistName={true}>
-                  {canEditProfile &&
-                    <div className="work-action-wrapper mb2">
-                      <Button type="hover-button" onClick={() => this.updateWork(work.id)}>
-                        <FontAwesomeIcon className="white" icon={faEdit} />
-                        <h4 className="ml2 white">Edit</h4>
+                <div>
+                {(!work.hidden || this.props.userType == "admin" || (this.props.user != null && (this.props.user.account_id == this.props.artist.account_id))) &&
```
Comment (by-co) [inline app/javascript/components/artists/ArtistProfile.jsx]: Nit: clean up the `this.props`  by destructuring, like we have done for `this.state` and `artist` on line 182!

File: app/javascript/components/artists/ArtistProfile.jsx [hunk 5]
```diff
@@ -119,6 +191,7 @@ class ArtistProfile extends React.Component {
     }
 
     const featured_work = works.find(work => work.id === artist.featured_work_id);
+    console.log(artist.locked_at)
```
Comment (by-co) [inline app/javascript/components/artists/ArtistProfile.jsx]: stray print

File: app/javascript/components/artists/AllArtists.jsx [hunk 6]
```diff
@@ -120,7 +163,24 @@ class AllArtists extends React.Component {
             </div>
             <div className="col-list-3">
               {artists.slice(artistStartIndex, artistEndIndex).map((artist, i) => {
-                return <ArtistColumnPanel key={i} artist={artist} />
+                return (
+                  <div>
+                    <ArtistColumnPanel key={i} artist={artist}>
+                      {artist.hidden == false && this.props.userType == "admin" &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.hideArtist(artist.id)}>
```
Comment (by-co) [inline app/javascript/components/artists/AllArtists.jsx]: Styling: leftmost button shouldn't have margin-left
```suggestion
                      <Button type="hover-button" onClick={() => this.hideArtist(artist.id)}>
```

File: app/javascript/components/artists/AllArtists.jsx [hunk 7]
```diff
@@ -120,7 +163,24 @@ class AllArtists extends React.Component {
...
+                        <FontAwesomeIcon className="white" icon={faTrash} />
+                        <h4 className="ml2 white">Hide</h4>
+                      </Button>
+                      }
+                      {artist.hidden == true && this.props.userType == "admin" &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.unHideArtist(artist.id)}>
```
Comment (by-co) [inline app/javascript/components/artists/AllArtists.jsx]:
```suggestion
                      <Button type="hover-button" onClick={() => this.unHideArtist(artist.id)}>
```

File: app/javascript/components/artists/ArtistProfile.jsx [hunk 8]
```diff
@@ -171,22 +256,42 @@ class ArtistProfile extends React.Component {
...
+                            <FontAwesomeIcon className="white" icon={faTrash} />
+                            <h4 className="ml2 white">Delete</h4>
+                          </Button>
+                        }
+                      </div>
+                    }
+                    {this.props.userType == "admin" &&
+                      <div>
+                      {work.hidden == false &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.hideWork(work.id)}>
```
Comment (by-co) [inline app/javascript/components/artists/ArtistProfile.jsx]: Left-most button again 
```suggestion
                      <Button type="hover-button" onClick={() => this.hideWork(work.id)}>
```

File: app/javascript/components/artists/ArtistProfile.jsx [hunk 9]
```diff
@@ -171,22 +256,42 @@ class ArtistProfile extends React.Component {
...
+                        <FontAwesomeIcon className="white" icon={faTrash} />
+                        <h4 className="ml2 white">Hide</h4>
                       </Button>
-                      {(work.availability == 'active') &&
-                      <Button className="ml2" type="hover-button" onClick={() => this.deleteWork(work.id)}>
+                      }
+                      {work.hidden == true &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.unHideWork(work.id)}>
```
Comment (by-co) [inline app/javascript/components/artists/ArtistProfile.jsx]:
```suggestion
                      <Button type="hover-button" onClick={() => this.unHideWork(work.id)}>
```

File: app/javascript/components/works/Works.jsx [hunk 10]
```diff
@@ -106,7 +147,22 @@ class Works extends React.Component {
           <div className="col-list-3">
             {works.slice(workStartIndex, workEndIndex).map((work, i) => {
               return (
-                <WorkColumnPanel key={i} work={work} />
+                <div>
+                  <WorkColumnPanel key={i} work={work}>
+                      {work.hidden == false && this.props.userType == "admin" &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.hideWork(work.id)}>
```
Comment (by-co) [inline app/javascript/components/works/Works.jsx]: Only one button here so we don't need margin-left~
```suggestion
                      <Button type="hover-button" onClick={() => this.hideWork(work.id)}>
```

File: app/javascript/components/works/Works.jsx [hunk 11]
```diff
@@ -106,7 +147,22 @@ class Works extends React.Component {
...
+                        <FontAwesomeIcon className="white" icon={faTrash} />
+                        <h4 className="ml2 white">Hide</h4>
+                      </Button>
+                      }
+                      {work.hidden == true && this.props.userType == "admin" &&
+                      <Button className="ml2" type="hover-button" onClick={() => this.unHideWork(work.id)}>
```
Comment (by-co) [inline app/javascript/components/works/Works.jsx]:
```suggestion
                      <Button type="hover-button" onClick={() => this.unHideWork(work.id)}>
```

File: config/routes.rb (hunk 1 above)
Comment (tommypoa) [reply config/routes.rb]: Put request is necessary since the controller method `lock_user` updates the `locked_at` field for devise to properly lock a user.
